src/go/pt-galera-log-explainer/internal.go

```diff
@@ -0,0 +1,218 @@
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+var logger = log.With().Str("component", "extractor").Logger()
+
+func init() {
+
+	if CLI.Since != nil {
+		logger = logger.With().Time("since", *CLI.Since).Logger()
+	}
+	if CLI.Until != nil {
+		logger = logger.With().Time("until", *CLI.Until).Logger()
+	}
+}
+
+var (
+	errDirectoriesUnsupported = errors.New("directories are not supported")
+)
+
+// timelineFromPaths takes every path, search them using a list of regexes
+// and organize them in a timeline that will be ready to aggregate or read
+func timelineFromPaths(paths []string, regexes types.RegexMap) (types.Timeline, error) {
+	timeline := make(types.Timeline)
+	found := false
+
+	compiledRegex := prepareGrepArgument(regexes)
+
+	for _, path := range paths {
+		osinfo, err := os.Stat(path)
+		if err != nil {
+			return nil, err
+		}
+		if osinfo.IsDir() {
+			return nil, errDirectoriesUnsupported
+		}
+
+		stdout := make(chan string)
+
+		go func() {
+			err := execGrepAndIterate(path, compiledRegex, stdout)
+			if err != nil {
+				logger.Error().Str("path", path).Err(err).Msg("execGrepAndIterate returned error")
+			}
+		}()
+
+		// it will iterate on stdout pipe results
+		localTimeline := iterateOnGrepResults(path, regexes, stdout)
+		if len(localTimeline) == 0 {
+			continue
+		}
+		found = true
+		logger.Debug().Str("path", path).Msg("finished searching")
+
+		// Why it should not just identify using the file path:
+		// so that we are able to merge files that belong to the same nodes
+		// we wouldn't want them to be shown as from different nodes
+		if CLI.PxcOperator {
+			timeline[path] = localTimeline
+		} else if CLI.MergeByDirectory {
+			timeline.MergeByDirectory(path, localTimeline)
+		} else {
+			timeline.MergeByIdentifier(localTimeline)
+		}
+	}
+	if !found {
+		return nil, errors.New("could not find data")
+	}
+	return timeline, nil
+}
+
+func prepareGrepArgument(regexes types.RegexMap) string {
+
+	regexToSendSlice := regexes.Compile()
+
+	grepRegex := "^"
+	if CLI.PxcOperator {
+		// special case
+		// I'm not adding pxcoperator map the same way others are used, because they do not have the same formats and same place
+		// it needs to be put on the front so that it's not 'merged' with the '{"log":"' json prefix
+		// this is to keep things as close as '^' as possible to keep doing prefix searches
+		grepRegex += "((" + strings.Join(regex.PXCOperatorMap.Compile(), "|") + ")|^{\"log\":\""
+		regexes.Merge(regex.PXCOperatorMap)
+	}
+	if CLI.Since != nil {
+		grepRegex += "(" + regex.BetweenDateRegex(CLI.Since, CLI.PxcOperator) + "|" + regex.NoDatesRegex(CLI.PxcOperator) + ")"
+	}
+	grepRegex += ".*"
+	grepRegex += "(" + strings.Join(regexToSendSlice, "|") + ")"
+	if CLI.PxcOperator {
+		grepRegex += ")"
+	}
+	logger.Debug().Str("grepArg", grepRegex).Msg("compiled grep arguments")
+	return grepRegex
+}
+
+func execGrepAndIterate(path, compiledRegex string, stdout chan<- string) error {
+
+	defer close(stdout)
```

Comment:
I suggest to close this channel outside this function.

Reply:
Done, it does make more sense